Description: Ensure armv7-a target complies with Debian armhf baseline
 Currently, Debian's armhf architecture baseline is armv7 with VFPv3-D16
 floating point, while NEON is not guaranteed[1].
 .
 However, upstream Clang assumes that features such as NEON are
 available on the armv7-a target. This patch ensures that the Debian
 armhf baseline is respected by disabling NEON generation by default on
 armv7-a.
 .
 [1]: https://wiki.debian.org/ArchitectureSpecificsMemo#armhf
Author: Sylvestre Ledru <sylvestre@debian.org>
Bug-Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=842142
Bug-Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=914268
Forwarded: not-needed
Last-Update: 2026-06-05
---
Index: llvm-toolchain-snapshot_21~++20250422120724+c1940cd0deef/llvm/include/llvm/TargetParser/ARMTargetParser.def
===================================================================
--- llvm-toolchain-snapshot_21~++20250422120724+c1940cd0deef.orig/llvm/include/llvm/TargetParser/ARMTargetParser.def
+++ llvm-toolchain-snapshot_21~++20250422120724+c1940cd0deef/llvm/include/llvm/TargetParser/ARMTargetParser.def
@@ -92,8 +92,8 @@ ARM_ARCH("armv6kz", ARMV6KZ, "6KZ", "+v6
          FK_VFPV2, (ARM::AEK_SEC | ARM::AEK_DSP))
 ARM_ARCH("armv6-m", ARMV6M, "6-M", "+v6m", ARMBuildAttrs::CPUArch::v6_M,
          FK_NONE, ARM::AEK_NONE)
-ARM_ARCH("armv7-a", ARMV7A, "7-A", "+v7", ARMBuildAttrs::CPUArch::v7, FK_NEON,
-         ARM::AEK_DSP)
+ARM_ARCH("armv7-a", ARMV7A, "7-A", "+v7", ARMBuildAttrs::CPUArch::v7,
+         FK_VFPV3_D16 /* Hard float */, ARM::AEK_DSP)
 ARM_ARCH("armv7ve", ARMV7VE, "7VE", "+v7ve", ARMBuildAttrs::CPUArch::v7,
          FK_NEON,
          (ARM::AEK_SEC | ARM::AEK_MP | ARM::AEK_VIRT | ARM::AEK_HWDIVARM |
Index: llvm-toolchain-snapshot_21~++20250422120724+c1940cd0deef/llvm/lib/Target/ARM/ARMArchitectures.td
===================================================================
--- llvm-toolchain-snapshot_21~++20250422120724+c1940cd0deef.orig/llvm/lib/Target/ARM/ARMArchitectures.td
+++ llvm-toolchain-snapshot_21~++20250422120724+c1940cd0deef/llvm/lib/Target/ARM/ARMArchitectures.td
@@ -42,7 +42,8 @@ def ARMv6sm   : Architecture<"armv6s-m",
                                                        FeatureStrictAlign]>;
 
 def ARMv7a    : Architecture<"armv7-a",   "ARMv7a",   [HasV7Ops,
-                                                       FeatureNEON,
+                                                       FeatureVFP3,
+                                                       FeatureVFP3_D16,
                                                        FeatureDB,
                                                        FeatureDSP,
                                                        FeatureAClass,
